gpii/node_modules/flowManager/src/UserLogonHandlers.js

```diff
@@ -0,0 +1,199 @@
+    fluid.log("proximity triggered by GPII key: ", gpiiKey);
+    // find currently logged in user:
+    var activeGpiiKeys = lifecycleManager.getActiveSessionGpiiKeys();
+
+    if (that.isBounce(gpiiKey, that)) {
+        fluid.log("Proximity trigger ignored due to bounce rules");
+        that.errorResponse("Proximity trigger ignored due to bounce rules. Please wait current logon change is complete", 429);
+        return;
+    }
+
+    // if no user is logged in, request a login
+    // if the login user is "reset", trigger the reset process via login with "reset" key
+    if (activeGpiiKeys.length === 0 || gpiiKey === "reset") {
+        var loginPromise = lifecycleManager.addToUserLogonRequestQueue({
+            gpiiKey: gpiiKey,
+            logonState: "login"
+        });
+        loginPromise.then(request.events.onSuccess.fire, request.events.onError.fire);
+        return;
+    }
+
+    // if user is already logged in, log them out. If it's a different user who is logged in
+    // log that user out before logging in with the new user
+    if (activeGpiiKeys.length === 1) {
+        var previousUser = activeGpiiKeys[0];
+        if (previousUser !== gpiiKey) {
+            var logoutPreviousUserPromise = lifecycleManager.addToUserLogonRequestQueue({
+                gpiiKey: previousUser,
+                logonState: "logout"
+            });
+            logoutPreviousUserPromise.then(function () {
+                var loginPromise = lifecycleManager.addToUserLogonRequestQueue({
+                    gpiiKey: gpiiKey,
+                    logonState: "login"
+                });
+                loginPromise.then(request.events.onSuccess.fire, request.events.onError.fire);
+            }, function () {
+                that.errorResponse("Unable to log previous user(" + previousUser + ") out so canceling login.");
+            });
+        } else {
+            var logoutPromise = lifecycleManager.addToUserLogonRequestQueue({
+                gpiiKey: previousUser,
+                logonState: "logout"
+            });
+            logoutPromise.then(request.events.onSuccess.fire, request.events.onError.fire);
+        }
+    }
+};
+
+// Request handler for /user/<myGpiiKey>/login
+fluid.defaults("gpii.flowManager.userLogin.handler", {
+    gradeNames: ["kettle.request.http", "gpii.flowManager.baseHandler"],
+    invokers: {
+        handleRequest: {
+            funcName: "gpii.flowManager.userLogin.handleRequest",
+            args: ["{request}.req.params.gpiiKey", "{request}", "{lifecycleManager}"]
+        }
+    }
+});
+
+/*
+ * Handler for URL: /user/<myGpiiKey>/login:
+ * * If no user is logged in, <myGpiiKey> will be logged in
+ * * If another user is logged in, nothing will happen
+ * * If GPII key is "reset", log out any already logged in user.
+ */
+gpii.flowManager.userLogin.handleRequest = function (gpiiKey, request, lifecycleManager) {
+    var loginPromise = lifecycleManager.addToUserLogonRequestQueue({
```

Comment:
These handlers are *very nearly* refactored compared to the proximity one, but they still need to be pushed the final step to where the login/logout action just consists of an invoker or event firing, to prepare for the day when we stop having HTTP endpoints at all. Also, note that the logout handler is identical except for the value of one string -> combine them.